lock-api/src/main/java/com/palantir/lock/client/LockTokenShare.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import com.google.common.base.Preconditions;
+import com.palantir.lock.v2.LockToken;
+
+final class LockTokenShare implements LockToken {
+    private final UUID requestId;
+    private final LockToken sharedLockToken;
+    private final ReferenceCounter referenceCounter;
+
+    private boolean unlocked;
+
+    private LockTokenShare(ReferenceCounter referenceCounter, LockToken token) {
+        this.referenceCounter = referenceCounter;
+        this.requestId = UUID.randomUUID();
+        this.sharedLockToken = token;
+        this.unlocked = false;
+    }
+
+    /**
+     * Returns a list of {@link LockTokenShare}, that are referencing to same LockToken.
+     *
+     * Share should be called only once on same {@link LockTokenShare} instance, guarantees do not hold between two
+     * different lists of {@link LockTokenShare} generated by successive share calls on same {@link LockToken}.
+     */
+    public static List<LockToken> share(LockToken token, int referenceCount) {
+        Preconditions.checkArgument(referenceCount > 0, "Reference count should be more than zero");
+        Preconditions.checkArgument(!(token instanceof LockTokenShare), "Can not share a shared lock token");
+        ReferenceCounter referenceCounter = new ReferenceCounter(referenceCount);
+        return IntStream.range(0, referenceCount)
+                .mapToObj(unused -> new LockTokenShare(referenceCounter, token))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Unlocks shared token on client side - does not guarantee underlying token to be unlocked on server side.
+     *
+     * @return referenced shared lock token if all lock token shares are unlocked; that is lock token on server side is
+     * good to be unlocked.
+     */
+    public synchronized Optional<LockToken> unlock() {
```

Comment:
There's some concurrency weirdness here. I don't think it's a bug, but I think it's probably worth paying attention to?

Basically, because the reference counter is locked differently to this class, two different classes can unmark concurrently and then both return the shared lock token.

This just means that sometimes, two people will try to unlock the same lock. This just leads to some weirdness wrt the lock contract; if I call unlock and it returns my lock token, it means that I had the lock for the whole period; here there's a way to end up with that incorrectly not being the case.

I'd have the unmark method return true if we actually unlocked the thing (and do that once)

```
if (!unlocked) {
    unlocked = true;
    return referenceCounter.unmark() ? Optional.of(sharedLockToken) : Optional.empty();
}
return Optional.empty();
```

Reply:
as we need to call refresh in any case to ensure that we were actually holding the locks before returning when unlock is called, we stick to contract in this implementation also I believe.
One problem with not returning the referenced token when it is already unlocked here is that, if you call unlock on lockTokenShare but fail to call timelock server to unlock, on the second call you may think the lock is actually unlocked, and just skip it.